tools/deploy_firmware.py

```diff
@@ -0,0 +1,287 @@
+        """
+        version = None
+        with pexpect.pxssh.pxssh(logfile=sys.stdout.buffer) as shell:
+            shell.login(self.name, self.username)
+            shell.sendline("cat /etc/prplwrt-version")
+            shell.expect("cat /etc/prplwrt-version")
+            shell.prompt()
+            version = shell.before.decode().splitlines()
+        return version
+
+    def read_artifacts_dir_version(self) -> List[str]:
+        """Reads the local prplwrt-version.
+
+        The version file is read from the artifacts (see `artifacts_dir`).
+
+        Returns
+        -------
+        List[str]
+            The content of the local prplwrt-version as a list of strings.
+        """
+        with open(os.path.join(self.artifacts_dir, "prplwrt-version")) as version_file:
+            return version_file.read().splitlines()
+
+    def needs_upgrade(self):
+        """Check if a device needs to be updated
+
+        The check is done by comparing prplwrt-version on the target
+        with a local one (see `read_artifacts_dir_version).
+        """
+        artifacts_dir_version = self.read_artifacts_dir_version()
+        remote_version = self.read_remote_version()
+        diff = difflib.unified_diff(artifacts_dir_version, remote_version,
+                                    fromfile='file1', tofile='file2')
+        diff_str = '\n'.join(diff)
+        if diff_str:
+            print(diff_str)
+        return bool(diff_str)
+
+    def reach(self, attempts: int = 5, wait: int = 5):
+        """Check if the device is reachable via SSH (and optionally try multiple times).
+
+        Parameters
+        ----------
+        attempts: int, optional
+            How many times we try before concluding that the device is unreachable.
+        wait: int, optional
+            The number of seconds to wait between attempts.
+
+        Returns
+        -------
+        bool
+            True if the device is reachable via SSH, false otherwise.
+        """
+        for _ in range(attempts):
+            try:
+                with pexpect.pxssh.pxssh() as shell:
+                    shell.login(self.name, self.username, login_timeout=5)
+                    return True
+            except pexpect.pxssh.ExceptionPxssh:
+                print("Waiting for the device to be reachable")
+                time.sleep(wait)
+        return False
+
+    def sysupgrade(self):
+        """Upgrade the device using sysupgrade."""
+        raise NotImplementedError
+
+    def upgrade_uboot(self):
+        """Upgrade the device through uboot."""
+        raise NotImplementedError
+
+
+class NetgearRax40(PrplwrtDevice):
+    """A Netgear rax40.
+
+    At the moment, the rax40 can only be updated through uboot using a tftp server.
+    """
+
+    tftp_dir = "/srv/tftp"
+    """The root directory of the tftp server. The image will be copied there"""
+
+    uboot_prompt = "GRX500 #"
+    """The u-boot prompt on the target."""
+
+    def upgrade_uboot(self):
```

Comment:
@LiorAmram will soon work on adding support for this script to upgrade UGW and RDKB.
Both UGW and prplWrt RAX40 are using the same uboot so upgrading a version is practially the same. UGW is however not a PrplwrtDevice so we won't be able to reuse this like this.

RAX40 is basically prplWrt with UGW BSP. The SoC is called GRX (we have GRX350 and GRX500 but the differences are negligible) - so its better to move the upgrade_uboot and sysupgrade to a different class. Perhaps `class device_operations` which will have abstract methods for `sysupgrade` and `uboot_upgrade`, and add it as a member of the prplwrt_device class?

Reply:
Since it is anyway work to rewrite the code to satisfy those future constraints, it is better to do that rewrite when we're actually going to do it, rather than trying to predict now what it should look like (in other words: merge as is, refactor later).

Reply:
Agh sorry that's exactly what you said.